### PR TITLE
Use cuSPARSE Generic API instead of older one documented to be removed

### PR DESCRIPTION
**Pull request by takagi**
Rel. #3513

Some cuSPARSE functions are announced to be removed in the next CUDA major release on its document. This PR fixes CuPy functions which use such older functions to use cuSPARSE Generic API instead.

**TODO**
- cupyx.scipy.sparse
  - cupyx/scipy/sparse/linalg/_solve.py
    - [x] fix spsolve_triangular to use spsm first (currently uses csrsm2)
    - [x] fix SuperLU.solve to use spsm first (curently uses csrsm2)
  - cupyx/scipy/sparse/_csc.py
    - [x] fix csc_matrix.\_\_mul__ to use spgemm
- cupy.cusparse
  - [x] update _available_cusparse_version
  - add spsm function
    - [x] CUDA
    - [x] HIP
- cupy_backends
  - add compat stubs
    - [x] add CUDA backend stubs for API to be removed in CUDA 12
  - add spsm bindings or stubs
    - [x] Cython
    - [x] cuda
    - [x] hip
    - [x] stub

**cuSPARSE API to be removed (32 cuSPARSE doc items)**

<details>

- cusparseAlgMode_t
- csrgemm2Info_t
- csrsm2Info_t
- csrsv2Info_t
- cusparseCreateCsrsv2Info
- cusparseDestroyCsrsv2Info
- cusparseCreateCsrsm2Info
- cusparseDestroyCsrsm2Info
- cusparseCreateCsrgemm2Info
- cusparseDestroyCsrgemm2Info
- cusparse\<t>axpyi
- cusparse\<t>gthr
- cusparse\<t>gthrz
- cusparse\<t>roti
- cusparse\<t>sctr
- cusparseCsrmvEx
- cusparse\<t>csrsv2_bufferSize
- cusparse\<t>csrsv2_analysis
- cusparse\<t>csrsv2_solve
- cusparseXcsrsv2_zeroPivot
- cusparse\<t>csrsm2_bufferSizeExt
- cusparse\<t>csrsm2_analysis
- cusparse\<t>csrsm2_solve
- cusparseXcsrsm2_zeroPivot
- cusparse\<t>gemmi
- cusparse\<t>csrgemm2
- cusparse\<t>csc2dense
- cusparse\<t>csr2dense
- cusparse\<t>dense2csc
- cusparse\<t>dense2csr
- cusparseSpMatSetStridedBatch
- cusparseConstrainedGeMM

</details>

### CODE REVIEW

**Comment by takagi** [comment]
I checked that cusparseSpSM_* was introduced in CUDA 11.3.1, and hipsparseSpSM_* in ROCm 5.0.0.

**Comment by takagi** [comment]
hipsparseSpSM_solve has an extra parameter `externalBuffer` that cusparseSpSM_solve does not have. I'll disable SpSM on HIP this time.

File: tests/cupyx_tests/scipy_tests/sparse_tests/test_linalg.py [hunk 1]
```diff
@@ -794,12 +794,21 @@ def _make_matrix(self, dtype, xp):
...
 
     def _test_spsolve_triangular(self, sp, a, b):
-        return sp.linalg.spsolve_triangular(a, b, lower=self.lower,
-                                            unit_diagonal=self.unit_diagonal)
+        def call_with_ignore_warning(f):
+            with warnings.catch_warnings():
+                warnings.simplefilter('ignore', sp.SparseEfficiencyWarning)
+                return f()
+        if sp is scipy.sparse:
+            call = call_with_ignore_warning
```
Comment (takagi) [inline tests/cupyx_tests/scipy_tests/sparse_tests/test_linalg.py:802]: Scipy only may warn SparseEfficiencyWarning

File: tests/cupyx_tests/scipy_tests/sparse_tests/test_linalg.py [hunk 2]
```diff
@@ -794,12 +794,21 @@ def _make_matrix(self, dtype, xp):
...
+        else:
+            call = lambda f: f()
+        return call(lambda: sp.linalg.spsolve_triangular(
+            a, b, lower=self.lower, unit_diagonal=self.unit_diagonal))
 
     @pytest.mark.parametrize('format', ['csr', 'csc', 'coo'])
     @testing.for_dtypes('fdFD')
-    @testing.numpy_cupy_allclose(rtol=1e-5, atol=1e-5, sp_name='sp')
+    @testing.numpy_cupy_allclose(
+        rtol=1e-5, atol=1e-5, sp_name='sp', contiguous_check=False)
```
Comment (takagi) [inline tests/cupyx_tests/scipy_tests/sparse_tests/test_linalg.py:811]: Add `contiguous_check=False` as spsm assumes f-order for the solution matrix `x` (on the other hand, csrsm2 overwrites b so the solution matrix has the same order with b)

File: tests/cupyx_tests/scipy_tests/sparse_tests/test_linalg.py [hunk 3]
```diff
@@ -794,12 +794,23 @@ def _make_matrix(self, dtype, xp):
...
     def _test_spsolve_triangular(self, sp, a, b):
-        return sp.linalg.spsolve_triangular(a, b, lower=self.lower,
-                                            unit_diagonal=self.unit_diagonal)
+        @contextlib.contextmanager
+        def ignore_warning():
+            with warnings.catch_warnings():
+                warnings.simplefilter('ignore', sp.SparseEfficiencyWarning)
+                yield
+        if sp is scipy.sparse:
+            ctx = ignore_warning
```
Comment (takagi) [inline tests/cupyx_tests/scipy_tests/sparse_tests/test_linalg.py:803]: SciPy only may warn SparseEfficiencyWarning.

**Comment by takagi** [comment]
@anaruse Is it safe to give the same matrix with B to the matrix C in cusparseSpSM_solve to overwrite B to store the solution?
https://docs.nvidia.com/cuda/cusparse/index.html#cusparse-generic-function-spsm

**Comment by takagi** [comment]
I'll kick full test after code unfrozen.

**Comment by takagi** [comment]
Let me run another full test after pre-review tests pass.

**Comment by kmaehashi** [comment]
Thanks for the hard work @takagi! Could you briefly check the performance difference before/after the change for these affected functions?

**Comment by takagi** [comment]
I've a comparison between `cupy.cusparse.csrsm2` and `cupy.cusparse.spsm`. The result weirdly says spsm is much slower than csrsm2 (and even than scipy's spsolve_triangular). I've also profiled the script with Python cProfile and found almost all of the spsm time is used by spSM_analysis (in cupy_backends).

m | n | nnz | scipy[ms] | csrsm[ms] | spsm[ms]
--|--|--|--|--|--
100 | 10 | 1133 | 1.399 | 1.047 | 3.165
1000 | 10 | 100871 | 15.201 | 1.836 | 16.932
10000 | 10 | 10008206 | 872.663 | 5.945 | 1225.227

<details>
<summary>click to see benchmark script</summary>

```py
import numpy
import scipy

import cupy
from cupy import testing
import cupyx
import cupyx.scipy.linalg
from cupyx import profiler

def make_sparse(m, density, dtype):
    a_shape = m, m
    a = testing.shaped_random(a_shape, numpy, dtype=dtype, scale=0.001)
    a -= numpy.diag(a)
    a_mask = testing.shaped_random(a_shape, numpy, dtype='f', scale=1)
    a[a_mask > density] = 0
    a_diag = numpy.random.uniform(-0.001, 0.001, size=(m,)).astype(dtype) + 1
    a_diag = numpy.diag(a_diag)
    a += a_diag
    a = numpy.tril(a)
    assert abs(numpy.linalg.det(a) - 1) < 0.1
    a = scipy.sparse.csr_matrix(a)
    a.sum_duplicates()
    return a

def make_dense(m, n, dtype):
    return numpy.array(numpy.random.uniform(0, 1, (m, n)).astype(dtype), order='f')

def verify(a_np, b_np, a_cp, b_cp):
    # verify with scipy
    x = scipy.sparse.linalg.spsolve_triangular(a_np, b_np,lower=True)
    testing.assert_allclose(a_np.todense() @ x, b_np, rtol=1e-5, atol=1e-5)

    # verify with csrsm2
    x = cupy.copy(b_cp)  # f-order
    assert x._f_contiguous
    cupy.cusparse.csrsm2(a_cp, x)
    testing.assert_allclose(a_cp.todense() @ x, b_cp, rtol=1e-5, atol=1e-5)

    # verify with spsm
    x = cupy.cusparse.spsm(a_cp, b_cp)
    testing.assert_allclose(a_cp.todense() @ x, b_cp, rtol=1e-5, atol=1e-5)

def get_time(perf):
    cpu_time = perf.cpu_times.mean()
    gpu_time = perf.gpu_times.mean()
    return max(cpu_time, gpu_time) * 1000  # ms

def benchmark(a_np, b_np, a_cp, b_cp):
    times = []

    # scipy spsolve_triangular
    perf = profiler.benchmark(scipy.sparse.linalg.spsolve_triangular, (a_np, b_np), n_repeat=10, name='scipy spsolve_triangular')
    times.append(get_time(perf))


    # csrsm2
    def _csrsm2(a, b):
        x = cupy.copy(b)  # f-order
        assert x._f_contiguous
        cupy.cusparse.csrsm2(a, x)
    perf = profiler.benchmark(_csrsm2, (a_cp, b_cp), n_repeat=10, name='cusparse.csrsm2')
    times.append(get_time(perf))

    # spsm
    perf = profiler.benchmark(cupy.cusparse.spsm, (a_cp, b_cp), n_repeat=10, name='cusparse.spsm')
    times.append(get_time(perf))

    return times

dtype = cupy.float32
n = 10
density = 0.2

print('m | n | nnz | scipy[ms] | csrsm[ms] | spsm[ms]')
print('--|--|--|--|--|--')
for m in [100, 1000, 10000]:
    a_np = make_sparse(m, density, dtype)
    b_np = make_dense(m, n, dtype)
    a_cp = cupyx.scipy.sparse.csr_matrix(a_np)
    b_cp = cupy.asarray(b_np, order='f')
    verify(a_np, b_np, a_cp, b_cp)
    ret = benchmark(a_np, b_np, a_cp, b_cp)
    print(f'{m} | {n} | {a_np.nnz} | {ret[0]:.3f} | {ret[1]:.3f} | {ret[2]:.3f}')
```
</details>

**Comment by takagi** [comment]
@anaruse Have you ever heard any cases like [this](https://github.com/cupy/cupy/pull/7052#issuecomment-1275512137) about the performance of cusparseSpSM? (I'm afraid that I don't immediately have naive C/C++ replicator)

**Comment by anaruse** [comment]
> @anaruse Have you ever heard any cases like [this](https://github.com/cupy/cupy/pull/7052#issuecomment-1275512137) about the performance of cusparseSpSM? (I'm afraid that I don't immediately have naive C/C++ replicator)

It is quite slow.. Please wait a little while while we investigate as well.

File: cupyx/cusparse.py [hunk 4]
```diff
@@ -105,11 +105,12 @@ def _call_cusparse(name, dtype, *args):
     'dense2csc': (8000, None),
     'dense2csr': (8000, None),
     'csr2csr_compress': (8000, None),
-    'csrsm2': (9020, None),
+    'csrsm2': (9020, 12000),
     'csrilu02': (8000, None),
     'denseToSparse': (11300, None),
     'sparseToDense': (11300, None),
     'spgemm': (11100, None),
+    'spsm': (11600, None),
```
Comment (kmaehashi) [inline cupyx/cusparse.py:113]: Could you add a comment describing why 11600 is used here?

File: cupyx/scipy/sparse/linalg/_solve.py [hunk 5]
```diff
@@ -390,6 +392,24 @@ def lsmr(A, b, x0=None, damp=0.0, atol=1e-6, btol=1e-6, conlim=1e8,
     return x, istop, itn, normr, normar, normA, condA, normx
 
 
+def _does_use_spsm(b):
```
Comment (kmaehashi) [inline cupyx/scipy/sparse/linalg/_solve.py:395]:
```suggestion
def _should_use_spsm(b):
```

File: cupyx/scipy/sparse/linalg/_solve.py [hunk 6]
```diff
@@ -390,6 +392,24 @@ def lsmr(A, b, x0=None, damp=0.0, atol=1e-6, btol=1e-6, conlim=1e8,
...
+    if not runtime.is_hip:
+        # cusparseSpSM in CUDA 11.5 and eariler produces wrong results with
```
Comment (kmaehashi) [inline cupyx/scipy/sparse/linalg/_solve.py:397]: Let's use spsm starting CUDA 12.0 as this PR needs to be backported and impacts existing users.

**Comment by takagi** [comment]
cusparseSpSM may take a long time to compute when the given input matrix has many non-zero elements. The amount of nnz at https://github.com/cupy/cupy/pull/7052#issuecomment-1275512137 is high enough (20%) compared to typical cases where cusparseSpSM is used. Thanks, @anaruse!

File: tests/cupyx_tests/test_cusparse.py [hunk 7]
```diff
@@ -1017,3 +1018,100 @@ def test_sparseToDense(self, dtype):
...
+        if not runtime.is_hip and _cusparse.get_build_version() < 11701:
+            # eariler than CUDA 11.6
+            if b_order == 'c':
+                pytest.skip("Older CUDA has a bug")
+        if runtime.is_hip:
+            if format == 'coo' or b_order == 'c':
+                pytest.skip('may be buggy or not supported')
+        a = self.sparse_matrix(self.a)
+        b = cupy.array(self.b, order=b_order)
+        c = cupy.cusparse.spsm(
```
Comment (anaruse) [inline tests/cupyx_tests/test_cusparse.py:1099]:
```suggestion
        c = cusparse.spsm(
```

**Review by asi1024 — APPROVED** [review]

**Comment by asi1024** [comment]
conflict: https://github.com/cupy/cupy/pull/7013

**Comment by takagi** [comment]
I'll make its backport

**Pull request merged**